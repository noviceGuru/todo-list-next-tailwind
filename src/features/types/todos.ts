export type Todo = {
	id: string | number
	task: string
}

export type Id = number | string

export type Task = {
	task: string
}

export type InputProps = {
	onChange: (e: React.ChangeEvent<HTMLInputElement>) => void
	value?: string
	defaultValue?: string
	placeholder?: string
}

export type ButtonTypes = "delete" | "add" | "save" | "discard" | "edit"

export type ButtonProps = {
	type: ButtonTypes
	onClick?: () => void
	disabled?: boolean
	testId?: string
}

export type PostAndPutResponse = {
	response: Todo
	isOk: boolean
}