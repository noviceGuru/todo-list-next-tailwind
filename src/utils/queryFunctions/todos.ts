import { Id, Todo } from "@/features/types/todos"

export async function getTodos(url: URL) {
	let response : {
		data : Todo[] | undefined
		error : Error | undefined
	} = {
		data : undefined,
		error : undefined
	}

	try {
		const res = await fetch(url, {cache: 'no-store'})
		if (!res.ok) {
			throw new Error('Failed to fetch todos')
		}
		const todos = (await res.json()) as Todo[]
		response.data = todos
	} catch (err) {
		response.error = err as Error
	}

	return response
}

export async function deleteATodo(url: URL, id: Id) {
	const res = await fetch(`${url}/${id}`, {
		method: "DELETE"
	})
	return res.ok
}

export async function putATodo(url: URL, body: Todo) {
	const res = await fetch(`${url}/${body.id}`, {
		headers: {
			"Content-Type": "application/json",
		},
		method: "PUT",
		body: JSON.stringify(body)
	})
	if (!res.ok) {
		throw new Error('Failed to fetch todos')
	}
	const todos = (await res.json()) as Todo
	return {
		response: todos,
		isOk: res.ok
	}
}

export async function postATodo(url: URL, body: Omit<Todo, "id">) {
	const res = await fetch(`${url}`, {
		method: "POST",
		headers: {
			"Content-Type": "application/json",
		},
		body: JSON.stringify(body)
	})
	if (!res.ok) {
		throw new Error('Failed to fetch todos')
	}
	const todos = (await res.json()) as Todo
	return {
		response: todos,
		isOk: res.ok
	}
}