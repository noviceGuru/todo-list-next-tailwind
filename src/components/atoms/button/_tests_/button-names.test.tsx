import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'

import Button from "../button"

import { ButtonTypes } from '@/features/types/todos'

const buttonsTypesAndExpectedTexts: ButtonTypes[] = ['add', 'delete', 'save', 'discard', 'edit']

buttonsTypesAndExpectedTexts.forEach(type => {
	it('renders the add button with the correct name and looks', () => {
		render(
			<Button
				type={type}
			/>)
		const button = screen.getByTestId(`button-${type}`)

		expect(button).toBeInTheDocument()
		expect(button).toMatchSnapshot()
	})
})