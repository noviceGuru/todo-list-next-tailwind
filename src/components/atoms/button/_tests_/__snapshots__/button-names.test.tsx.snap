// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the add button with the correct name and looks 1`] = `
<button
  class="p-2 rounded-full w-10 h-10 disabled:opacity-50 disabled:cursor-not-allowed
			 bg-blue-300 hover:bg-blue-400"
  data-testid="button-add"
  type="button"
>
  <img
    alt="add-button"
    class="w-full"
    data-nimg="1"
    decoding="async"
    height="40"
    loading="lazy"
    src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
    srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
    style="color: transparent;"
    width="40"
  />
</button>
`;

exports[`renders the add button with the correct name and looks 2`] = `
<button
  class="p-2 rounded-full w-10 h-10 disabled:opacity-50 disabled:cursor-not-allowed
			 bg-red-400 border-red-400 hover:bg-red-600"
  data-testid="button-delete"
  type="button"
>
  <img
    alt="delete-button"
    class="w-full"
    data-nimg="1"
    decoding="async"
    height="40"
    loading="lazy"
    src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
    srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
    style="color: transparent;"
    width="40"
  />
</button>
`;

exports[`renders the add button with the correct name and looks 3`] = `
<button
  class="p-2 rounded-full w-10 h-10 disabled:opacity-50 disabled:cursor-not-allowed
			 bg-blue-300 hover:bg-blue-400"
  data-testid="button-save"
  type="button"
>
  <img
    alt="save-button"
    class="w-full"
    data-nimg="1"
    decoding="async"
    height="40"
    loading="lazy"
    src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
    srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
    style="color: transparent;"
    width="40"
  />
</button>
`;

exports[`renders the add button with the correct name and looks 4`] = `
<button
  class="p-2 rounded-full w-10 h-10 disabled:opacity-50 disabled:cursor-not-allowed
			 bg-blue-300 hover:bg-blue-400"
  data-testid="button-discard"
  type="button"
>
  <img
    alt="discard-button"
    class="w-full"
    data-nimg="1"
    decoding="async"
    height="40"
    loading="lazy"
    src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
    srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
    style="color: transparent;"
    width="40"
  />
</button>
`;

exports[`renders the add button with the correct name and looks 5`] = `
<button
  class="p-2 rounded-full w-10 h-10 disabled:opacity-50 disabled:cursor-not-allowed
			 bg-blue-300 hover:bg-blue-400"
  data-testid="button-edit"
  type="button"
>
  <img
    alt="edit-button"
    class="w-full"
    data-nimg="1"
    decoding="async"
    height="40"
    loading="lazy"
    src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
    srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
    style="color: transparent;"
    width="40"
  />
</button>
`;
